--- ../src-base/minecraft/net/minecraft/entity/ai/EntityAITasks.java
+++ ../src-work/minecraft/net/minecraft/entity/ai/EntityAITasks.java
@@ -1,5 +1,6 @@
 package net.minecraft.entity.ai;
 
+import carpet.helpers.AIHelper;
 import com.google.common.collect.Sets;
 import java.util.Iterator;
 import java.util.Set;
@@ -12,7 +13,7 @@
 {
     private static final Logger field_151506_a = LogManager.getLogger();
     private final Set<EntityAITasks.EntityAITaskEntry> field_75782_a = Sets.<EntityAITasks.EntityAITaskEntry>newLinkedHashSet();
-    private final Set<EntityAITasks.EntityAITaskEntry> field_75780_b = Sets.<EntityAITasks.EntityAITaskEntry>newLinkedHashSet();
+    public final Set<EntityAITasks.EntityAITaskEntry> field_75780_b = Sets.<EntityAITasks.EntityAITaskEntry>newLinkedHashSet(); // CM public
     private final Profiler field_75781_c;
     private int field_75778_d;
     private int field_75779_e = 3;
@@ -60,6 +61,8 @@
         {
             for (EntityAITasks.EntityAITaskEntry entityaitasks$entityaitaskentry : this.field_75782_a)
             {
+                if(carpet.carpetclient.CarpetClientChunkLogger.logger.enabled)
+                    carpet.carpetclient.CarpetClientChunkLogger.setReason(AIHelper.getInfo(this, entityaitasks$entityaitaskentry.field_75733_a));
                 if (entityaitasks$entityaitaskentry.field_188524_c)
                 {
                     if (!this.func_75775_b(entityaitasks$entityaitaskentry) || !this.func_75773_a(entityaitasks$entityaitaskentry))
@@ -67,6 +70,7 @@
                         entityaitasks$entityaitaskentry.field_188524_c = false;
                         entityaitasks$entityaitaskentry.field_75733_a.func_75251_c();
                         this.field_75780_b.remove(entityaitasks$entityaitaskentry);
+                        AIHelper.update(this);
                     }
                 }
                 else if (this.func_75775_b(entityaitasks$entityaitaskentry) && entityaitasks$entityaitaskentry.field_75733_a.func_75250_a())
@@ -74,8 +78,10 @@
                     entityaitasks$entityaitaskentry.field_188524_c = true;
                     entityaitasks$entityaitaskentry.field_75733_a.func_75249_e();
                     this.field_75780_b.add(entityaitasks$entityaitaskentry);
+                    AIHelper.update(this);
                 }
             }
+            carpet.carpetclient.CarpetClientChunkLogger.resetReason();
         }
         else
         {
@@ -102,8 +108,12 @@
 
             for (EntityAITasks.EntityAITaskEntry entityaitasks$entityaitaskentry2 : this.field_75780_b)
             {
+                if(carpet.carpetclient.CarpetClientChunkLogger.logger.enabled)
+                    carpet.carpetclient.CarpetClientChunkLogger.setReason(AIHelper.getInfo(this, entityaitasks$entityaitaskentry2.field_75733_a));
+                
                 entityaitasks$entityaitaskentry2.field_75733_a.func_75246_d();
             }
+            carpet.carpetclient.CarpetClientChunkLogger.resetReason();
 
             this.field_75781_c.func_76319_b();
         }
@@ -180,7 +190,7 @@
         }
     }
 
-    class EntityAITaskEntry
+    public class EntityAITaskEntry // CM public
     {
         public final EntityAIBase field_75733_a;
         public final int field_75731_b;
